In [1]:
import pandas as pd

from sklearn.datasets import fetch_california_housing

# Load the housing dataset
housing = fetch_california_housing()

In [ ]:
#Create dataframe for features (X) and series for labels (y)

X = pd.DataFrame(housing.data, columns=housing.feature_names) 
y = pd.Series(housing.target, name='med_house_value')

In [27]:
#Preview first 5 rows of X
X.head(5)

#Feature names and missing value counts for X
X.isnull().sum()

#Summary statistics for X
X.describe().round(2)

#Preview first 5 entries in y
y.head(5)

#Missing value sum of y
y.isnull().sum()

#Summary statistics for y
y.describe().round(2)

count    20640.00
mean         2.07
std          1.15
min          0.15
25%          1.20
50%          1.80
75%          2.65
max          5.00
Name: med_house_value, dtype: float64

In [ ]:
# Import linear regression class
from sklearn.linear_model import LinearRegression

# Import model splitting class
from sklearn.model_selection import train_test_split

# Split the data (80% training, 20% testing)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 42)

# Initialize and train the linear regression model
lin_reg = LinearRegression()
lin_reg.fit(X_train, y_train)

# Make predictions on the test set
y_pred = lin_reg.predict(X_test)

In [28]:
from sklearn.metrics import mean_squared_error, root_mean_squared_error, r2_score

# Evaluate model performance
mse = mean_squared_error(y_test, y_pred)
rmse = root_mean_squared_error(y_test, y_pred)

r2 = r2_score(y_test, y_pred)

print("Housing Data Linear Model:")
print(f"Mean Squared Error: {mse:.2f}")
print(f"Root Squared Error: {rmse:.2f}")
print(f"R² Score: {r2:.2f}")

Housing Data Linear Model:
Mean Squared Error: 0.56
Root Squared Error: 0.75
R² Score: 0.58


In [32]:
# View our model's coefficients
print("Model Coefficients:")
print(pd.Series(lin_reg.coef_,
                index=X.columns))
print("\nModel Intercept:")
print(pd.Series(lin_reg.intercept_))

Model Coefficients:
MedInc        0.448675
HouseAge      0.009724
AveRooms     -0.123323
AveBedrms     0.783145
Population   -0.000002
AveOccup     -0.003526
Latitude     -0.419792
Longitude    -0.433708
dtype: float64

Model Intercept:
0   -37.023278
dtype: float64


# Evaluating our Model

1. **R-Squared** The R-squared value for this model is 0.58, meaning that 58% of the variance in house value can be explained by a linear relationship with the features. This is not extremely strong, but not very weak either. This model has moderate strength.

2. **Features** Based on the model's coefficients, the 3 features that seem to have the strongest impact on predictions are MedInc, AveBedrms, and Longitude, because they have the 3 largest absolute value coefficients. However, this could be skewed or not entirely accurate because the data is unscaled.

3. **Distance** For this model, the MSE is 0.56, meaning that the average squared distance between actual and predicted values is 0.56. The RMSE is 0.75, meaning that the square root of MSE, a measure of distance in the same units as the target, is 0.75. Because the mean of the target is only around 2, the RMSE being 0.75 is relatively large, indicating that our predicted values may not match our actual values very closely.

### Feature Selection

Based on the above, a smaller model with only 3 features would perhaps make sense to choose the features that seemed to have the largest impact on predictions according to coefficients. As such, I choose to use MedInc, AveBedrms, and Longitude to test a smaller model

In [ ]:
#Create smaller X with selected features
X_2 = X[['MedInc', 'AveBedrms', 'Longitude']]

#Split test and train again (80% train, 20% test)
X2_train, X2_test, y2_train, y2_test = train_test_split(X_2, y, test_size = 0.2, random_state = 42)

# Initialize and train new linear regression model
lin_reg2 = LinearRegression()
lin_reg2.fit(X2_train, y2_train)

# Make new predictions on the test set
y2_pred = lin_reg2.predict(X2_test)

In [ ]:
# Evaluate new model performance
mse_2 = mean_squared_error(y2_test, y2_pred)
rmse_2 = root_mean_squared_error(y2_test, y2_pred)

r2_2 = r2_score(y2_test, y2_pred)

print("Reduced Data Model:")
print(f"Mean Squared Error: {mse_2:.2f}")
print(f"Root Squared Error: {rmse_2:.2f}")
print(f"R² Score: {r2_2:.2f}")

Unscaled Data Model:
Mean Squared Error: 0.71
Root Squared Error: 0.84
R² Score: 0.46


In [34]:
# View new model's coefficients
print("Reduced Model Coefficients:")
print(pd.Series(lin_reg2.coef_,
                index=X_2.columns))
print("\nReduced Model Intercept:")
print(pd.Series(lin_reg2.intercept_))

Reduced Model Coefficients:
MedInc       0.418949
AveBedrms   -0.001910
Longitude   -0.019934
dtype: float64

Reduced Model Intercept:
0   -1.935608
dtype: float64


# Evaluation of Reduced Model

The reduced model performs worse than the full model. It has a higher MSE (0.71), RMSE (0.84), and a lower R-squared (0.46). This means that the reduced model explains less of the variance in the target, and it makes predictions with less accuracy. As such, I would not use this reduced model in practice, because, even though it reduces the number of predictors, which can be good, it loses enough strength that I would consider not worth it. However, it could be worthwhile to try other reduced models to see if one might lose less of the predictive power of the model. 